#1300 Supporting regional clusters for node pools

darrenhaken opened the pull request:
This is a WIP for creating node pools with regional clusters.

darrenhaken commented:
#1300 related

darrenhaken commented:
#1300

danawillow commented:
Hey @darrenhaken, are you looking for preliminary feedback or did you want us to wait until it's no longer WIP?

darrenhaken commented:
@kri5 I tested using the v1Beta1 API to create zonal clusters and it seemed to work fine (the tests passed). With that, I felt it wouldn't impact TF users which version of the API to use and it simplifies the codebase.

Is there some context I may have missed?

kri5 commented:
If you look at the cluster_container resource, they choose to maintain both version of the API, by adding a "region" field, and handling location by generating it from region or zone for the beta api only.

As the beta API are provided as-is, with no SLA, i think we should always have a way to call the stable API, which is provided with SLA.

darrenhaken commented:
@kri5 So Justification over the SLA makes sense, I can look to add back in the logic to use the different API version when a zone is provided.

I'm not completely sold on parsing zone to the location in the code, I'd prefer to add `location` as a new param to the TF. `location` is part of the API and `zone` is in fact deprecated so encouraging TF users to migrate to `location` felt sensible.

Is there a way in TF to mark a field as deprecated?

Thoughts on general approach?

kri5 commented:
My point here is, that until this API is GA, zone is the proper way of provision it, and location should be marked Beta, but no the other way around, ie. marking zone as deprecated, which is not yet the case.

darrenhaken commented:
Sure, so how about I add `location` and in the docs I say that's a beta param for regional clusters?

kri5 commented:
I'm not against adding location to the nodepool ressource, but the main issue for me is the non uniformity of having location in the node pool resource, and region in the container cluster resource.

That might confuse users, as theses two resources are usually used together.

What do you think?

darrenhaken commented:
@kri5 do you think `region` for conformity then? The negative to this is it looks like the underlying SDK  uses the location language so it is a deviation.

kri5 commented:
@darrenhaken I do not know what would be the more appropriate.

If you really want to reflect the changes of API in the beta, maybe the right thing would be to change region to location in the cluster resource?

darrenhaken commented:
The other option is extract the location from the cluster resource using
the cluster name when Zone is not provided and assume it's regional

The advantage to this is that the node pool needs to be in the same region
as the cluster so it maintains this coupling.

No idea if it's possible to extract from another resource but I can't see
why it wouldn't.

On Thu, 12 Apr 2018, 18:52 Kri5, <notifications@github.com> wrote:

> @darrenhaken <https://github.com/darrenhaken> I do not know what would be
> the more appropriate.
>
> If you really want to reflect the changes of API in the beta, maybe the
> right thing would be to change region to location in the cluster resource?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-380890784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me1rPM9oXGkJ_gZdpZLKIrWXuV4Jyks5tn5RxgaJpZM4TQb8T>
> .
>

danawillow commented:
Hey all, just piping in on the conversation (without looking at the code)- I'd prefer that node pool matches cluster with having region and zone separately instead of location. Both approaches (region/zone vs location) have their merits, but from talking to my team and the original PR author for regional clusters, we came to the decision to do region/zone. Whether it's really the best way or not is still up for debate, but for consistency the two resources should definitely match.

As for the beta/v1 APIs thing, it actually doesn't matter which endpoint of the API you call (this is new information my team and I learned not too long ago), but for code review purposes it's much easier if a given PR does exactly one thing, and so I'd recommend keeping the logic as-is. We have some changes coming down the road with how we handle beta features (part of our codegen efforts), so stay tuned for that as well.

darrenhaken commented:
Thanks for the input Dana, I'll change to region as advised.

I had removed the switch between code versions because it simplifies the
changes I need to make. Would it be ok to stick with that?

On Thu, 12 Apr 2018, 19:20 Dana Hoffman, <notifications@github.com> wrote:

> Hey all, just piping in on the conversation (without looking at the code)-
> I'd prefer that node pool matches cluster with having region and zone
> separately instead of location. Both approaches (region/zone vs location)
> have their merits, but from talking to my team and the original PR author
> for regional clusters, we came to the decision to do region/zone. Whether
> it's really the best way or not is still up for debate, but for consistency
> the two resources should definitely match.
>
> As for the beta/v1 APIs thing, it actually doesn't matter which endpoint
> of the API you call (this is new information my team and I learned not too
> long ago), but for code review purposes it's much easier if a given PR does
> exactly one thing, and so I'd recommend keeping the logic as-is. We have
> some changes coming down the road with how we handle beta features (part of
> our codegen efforts), so stay tuned for that as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-380898842>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me1qI7SQtbkFO-jsE0xsDuGo9qGpDks5tn5rdgaJpZM4TQb8T>
> .
>

danawillow commented:
For now let's say tentatively yes

darrenhaken commented:
Ok, I should have a working PR tomorrow so I'll message for review. Thanks
for looping back

On Thu, 12 Apr 2018, 19:29 Dana Hoffman, <notifications@github.com> wrote:

> For now let's say tentatively yes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-380901594>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me0e9h3x90Ic7wwNnJI8JwY2gbMvxks5tn5z7gaJpZM4TQb8T>
> .
>

darrenhaken commented:
@danawillow I believe this PR is code complete now. 

@Stono is going to test this as well with the existing TF he uses

danawillow commented on google/utils.go:
```diff
@@ -48,6 +48,17 @@ func getZone(d TerraformResourceData, config *Config) (string, error) {
 	return GetResourceNameFromSelfLink(res.(string)), nil
 }
 
+func getParent(d TerraformResourceData, config *Config) (string, error) {
```
I don't think this gets used anywhere

danawillow commented on website/docs/r/container_node_pool.html.markdown:
```diff
@@ -89,11 +89,34 @@ resource "google_container_cluster" "primary" {
+
+resource "google_container_node_pool" "regional-np" {
+  name       = "my-node-pool"
+  region       = "us-central1"
```
nit: `=` alignment

darrenhaken replied:
I actually appreciate all your attention to detail. This TF code is very vast! 👍

danawillow commented on google/resource_container_node_pool.go:
```diff
@@ -59,7 +50,12 @@ func resourceContainerNodePool() *schema.Resource {
 				"cluster": &schema.Schema{
 					Type:     schema.TypeString,
-					Required: true,
+					Optional: true,
```
Why is `Cluster` optional now?

darrenhaken replied:
I can revert this back, my original design was to expose the `parent` parameter and expect the TF user to construct the string to match. As we have gone with `region` this can definitely be reverted.

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -155,11 +151,81 @@ var schemaNodePool = map[string]*schema.Schema{
-	containerApiVersion := getContainerApiVersion(d, ContainerNodePoolBaseApiVersion, ContainerNodePoolVersionedFeatures)
-	config := meta.(*Config)
+func getNodePoolRegion(d TerraformResourceData, config *Config) (string, error) {
+	res, ok := d.GetOk("region")
```
@danawillow I'm not too sure what you mean. Are you saying an error would occur when the node pool is nested?

danawillow replied:
Sure, happy to elaborate. You can define a cluster with a node pool resource inside of it:

```
resource "google_container_cluster" "with_node_pool" {
	name = "cluster"
	zone = "us-central1-a"

	node_pool {
		name               = "np"
		initial_node_count = 2
	}
}
```

If you check out the code for `google_container_cluster`, you can see where it calls into some of the functions inside `resource_container_node_pool.go`. If you do a d.Get() from there, you're still doing it on the cluster resource, and so it'll use the cluster schema rather than the node pool schema. That's why a lot of the code inside of the node pool resource looks like `d.Get(prefix + "something")`: that lets the cluster and node pool resource share the code.

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -155,11 +151,81 @@ var schemaNodePool = map[string]*schema.Schema{
+	return GetResourceNameFromSelfLink(res.(string)), nil
+}
+
+func generateLocation(d TerraformResourceData, config *Config) (string, error) {
```
Yep makes sense.

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -169,97 +235,53 @@ func resourceContainerNodePoolCreate(d *schema.ResourceData, meta interface{}) e
-	d.SetId(fmt.Sprintf("%s/%s/%s", zone, cluster, nodePool.Name))
+	waitErr := containerBetaOperationWait(config,
+		operation, nodePoolInfo.project,
+		nodePoolInfo.location, "timeout creating GKE NodePool", timeoutInMinutes, 3)
```
Sure changed this.

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -169,97 +235,53 @@ func resourceContainerNodePoolCreate(d *schema.ResourceData, meta interface{}) e
 		return waitErr
 	}
 
+	d.SetId(fmt.Sprintf("%s/%s/%s", nodePoolInfo.location, nodePoolInfo.cluster, nodePool.Name))
```
Moved it back up, I didn't realise it could have such an impact

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -327,6 +351,11 @@ func resourceContainerNodePoolDelete(d *schema.ResourceData, meta interface{}) e
 	return nil
 }
 
+func getNodePoolName(id string) string {
```
I've moved it back to the bottom of the file. I had originally deleted the function as it wasn't needed but of course it was in the end :)

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -335,15 +364,26 @@ func resourceContainerNodePoolExists(d *schema.ResourceData, meta interface{}) (
 
-	_, err = config.clientContainer.Projects.Zones.Clusters.NodePools.Get(
-		project, zone, cluster, name).Do()
+	fullyQualifiedName := fmt.Sprintf("projects/%s/locations/%s/clusters/%s/nodePools/%s",
```
I think you can use the nodepoolinfo so i'll switch to that, good spot.

darrenhaken commented on google/resource_container_node_pool.go:
```diff
@@ -435,7 +481,7 @@ func flattenNodePool(d *schema.ResourceData, config *Config, np *containerBeta.N
 			return nil, fmt.Errorf("Error reading instance group manage URL '%q'", url)
 		}
-		igm, err := config.clientCompute.InstanceGroupManagers.Get(matches[1], matches[2], matches[3]).Do()
+		igm, err := config.clientComputeBeta.InstanceGroupManagers.Get(matches[1], matches[2], matches[3]).Do()
```
Thanks

darrenhaken commented:
@danawillow I have resolved most of the things you asked me to look at. There's a couple of things I wasn't clear on, can you get back to me and I'll try and fix them today.

As an aside, do you know if there's any appetite to explore unit tests or writing some tests against a stubbed interface? It feels like we could shorten the feedback loop on errors.

I am also having a couple of tests failing (one of them does on master too because of a K8 version which is no longer provided). Got any time to give some input if I share the test log?

darrenhaken commented:
Thought I'd mention I am seeing intermittent raise conditions, an example is deleting a cluster when it's got another operation being applied to it.

Got any ideas looking at the code? I'm not sure how reliable the tests generally are, I noticed some of them randomly failing on master when I tried that too.

darrenhaken commented:
Thanks that helps a lot, I'll try and take a look at it. Any suggestions
for what to change?

Do you have any ideas what may be causing failures to occur? The errors I'm
seeing are around attempting to delete resources in use. I can provide the
errors

On Mon, 16 Apr 2018, 20:10 Dana Hoffman, <notifications@github.com> wrote:

> *@danawillow* commented on this pull request.
> ------------------------------
>
> In google/resource_container_node_pool.go
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#discussion_r181853442>
> :
>
> > @@ -155,11 +151,81 @@ var schemaNodePool = map[string]*schema.Schema{
>  	},
>  }
>
> -func resourceContainerNodePoolCreate(d *schema.ResourceData, meta interface{}) error {
> -	containerApiVersion := getContainerApiVersion(d, ContainerNodePoolBaseApiVersion, ContainerNodePoolVersionedFeatures)
> -	config := meta.(*Config)
> +func getNodePoolRegion(d TerraformResourceData, config *Config) (string, error) {
> +	res, ok := d.GetOk("region")
>
> Sure, happy to elaborate. You can define a cluster with a node pool
> resource inside of it:
>
> resource "google_container_cluster" "with_node_pool" {
> 	name = "cluster"
> 	zone = "us-central1-a"
>
> 	node_pool {
> 		name               = "np"
> 		initial_node_count = 2
> 	}
> }
>
> If you check out the code for google_container_cluster, you can see where
> it calls into some of the functions inside resource_container_node_pool.go.
> If you do a d.Get() from there, you're still doing it on the cluster
> resource, and so it'll use the cluster schema rather than the node pool
> schema. That's why a lot of the code inside of the node pool resource looks
> like d.Get(prefix + "something"): that lets the cluster and node pool
> resource share the code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#discussion_r181853442>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me6VU1-CkXAWUtFqnA8zcBIE35FuQks5tpOy1gaJpZM4TQb8T>
> .
>

danawillow commented:
That particular one is considered in the code to be a retryable error, so it's possible we just need to increase the amount of time we retry for: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_container_cluster.go#L700.

Right now the only tests that are failing in CI are:
TestAccContainerCluster_withIPAllocationPolicy
TestAccContainerCluster_withNetworkPolicyEnabled
TestAccContainerNodePool_withGPU

The others are mostly reliable- we get some flakes once in a while but not all the time. If you want to post your debug logs for tests with other errors, I can probably take a look and see if there's anything obvious.

For tests against a stubbed interface, it's definitely something we've thought about, but it would be a pretty big undertaking so no concrete plans on our end at the moment. If you have ideas of how it could be done elegantly without a ton of engineering effort and with the same correctness guarantees as our current acceptance tests, I'm all ears :) In the meantime, you're welcome to write as many unit tests as you want! We have a bunch sprinkled around the code base already.

I'm only sort of working today, so I probably won't have a chance to do a thorough review, but looks like Nathan requested himself so you won't be left hanging.

ndmckinley commented:
Yep, I'll bring this to conclusion - let me know when you're ready.  :)

darrenhaken commented:
@danawillow I won't get time to look at fixing things today but tomorrow I should have time.

I have been experiencing the following error with the new regional test:

```
-----------------------------------------------------
2018/04/16 16:29:01 [WARN] WaitForState timeout after 2m0s
2018/04/16 16:29:01 [WARN] WaitForState starting 30s refresh grace period
2018/04/16 16:29:01 [ERROR] root: eval: *terraform.EvalRefresh, err: google_container_cluster.cluster: Error reading Container Cluster "tf-nodepool-test-ftvqrx9vxt": Cluster "tf-nodepool-test-ftvqrx9vxt" has status "RECONCILING" with message ""
2018/04/16 16:29:01 [ERROR] root: eval: *terraform.EvalSequence, err: google_container_cluster.cluster: Error reading Container Cluster "tf-nodepool-test-ftvqrx9vxt": Cluster "tf-nodepool-test-ftvqrx9vxt" has status "RECONCILING" with message ""
--- FAIL: TestAccContainerNodePool_regionalClusters (333.85s)
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error refreshing: 1 error(s) occurred:

		* google_container_cluster.cluster: google_container_cluster.cluster: Error reading Container Cluster "tf-nodepool-test-ftvqrx9vxt": Cluster "tf-nodepool-test-ftvqrx9vxt" has status "RECONCILING" with message ""

		State: <nil>
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	333.888s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-google/scripts	(cached) [no tests to run]
make: *** [testacc] Error 1
```

I had assumed it was related to some of the code changes you asked me to make but I reverted back to #6dd0ae3 and the same issue happened.

I think initially I'll focus on adding unit tests to the codebase, after all, they are the cheapest to write and yield the highest value. For example, writing unit tests around some of the utility functions feel like a simple task. I'd prefer to do that in another improvement PR though to move this along.

I'm not 100% sure what the fix is for the nested node pools but I will take a look tomorrow when I have time to read it properly.

@ndmckinley any help you can do to move this along is appreciated! Where I work are incredibly keen to get regional node pools so we can build our new regional clusters 👍

danawillow commented:
Yup, that's the error I think can be fixed by increasing the retry time in the code snippet I linked to.

darrenhaken commented:
Ah awesome I'll take a look then!


On Mon, 16 Apr 2018, 21:25 Dana Hoffman, <notifications@github.com> wrote:

> Yup, that's the error I think can be fixed by increasing the retry time in
> the code snippet I linked to.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-381737404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me5Q968iS1kfUfeYaw_ZckJXic6maks5tpP4lgaJpZM4TQb8T>
> .
>

darrenhaken commented:
@ndmckinley quick update on this. As per @danawillow advice, I've added the retry handler to the node pool creation. 

I have also introduced it into the delete function as well, I was experiencing errors from the GCloud API that Node pools were trying to be deleted in the acceptance tests whilst the cluster was busy.

One caveat with the deletion retry; it will attempt 3 retries regardless of the error which comes back. Is this OK or is there a more specific error condition?

I'm about to start investigating the other thing @danawillow raised which is handling nested `node_pool` within a cluster resource.

ndmckinley commented:
Great - I'll dig through those comments and mark anything other than nested node pools that I think still needs to be resolved.

Stono commented:
I spoke to Darren in the office earlier and at the time it wasn't clear
what we need to do to resolve the nested pool issue despite the earlier
comments.

Is it just to do with scoping of GetOk?

On Tue, 17 Apr 2018, 7:39 pm Nathan McKinley, <notifications@github.com>
wrote:

> Great - I'll dig through those comments and mark anything other than
> nested node pools that I think still needs to be resolved.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-382098455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaviYMlqXbJ2VlGvL-PFDLdpjBTc3Xnks5tpjb2gaJpZM4TQb8T>
> .
>

darrenhaken commented:
Thought I'd give a quick update @ndmckinley as I've pushed another commit.

I took a look through the code on master and the only place I could see a reference to the `prefix` is with the update of the cluster. I have now rolled back in the `nodePoolUpdate` function to:
* Use the cluster name passed in - this is to work with both node pools and node pools called from the cluster resource
* Use the prefix when handling the node pool ID.

I have looked back at the comments @danawillow had given but I have since removed the function that the original comments were left. I have actually started sharing the regional logic between `cluster` and `node_pool` which can be found in `regional_utils.go`. 

Could you take a look and see if it's still an issue?

I will also write an acceptance test tomorrow using a nested node pool and see how it all behaves.

ndmckinley commented on google/resource_container_node_pool_test.go:
```diff
@@ -390,6 +405,29 @@ func testAccCheckContainerNodePoolDestroy(s *terraform.State) error {
 	return nil
 }
 
+func TestAccContainerNodePool_regionalClusters(t *testing.T) {
```
This is unusual, but this resource is unusually complicated - could you also find a way to test that updates work correctly?  Maybe add another similar config and a test step to update from this config to that one?

ndmckinley replied:
I am also concerned about nested node pools - I see that you have a plan to write an acceptance test for those tomorrow, which is great.

darrenhaken replied:
What do you mean by unusual? It felt pretty complicated to me too.

ndmckinley replied:
It's not an important note - I'm just saying that this resource contains more twists and turns than most - most resources have very clean "create, update, delete" cycles where each step makes exactly one (simple) request, but the nature of these container node pools means that they have lots of special cases.

ndmckinley replied:
Oh!  Sorry, to clarify: my *request* is unusual - normally we would be very happy with this test as it stands, but since this resource is so much more complicated, I'd like to see some other testing.

darrenhaken replied:
Ah I see! Yes I'm fine to add more tests :-) 

Two more cases make sense to me:
- Doing an update (if this is possible)
- nested node pool

Thoughts?

ndmckinley commented:
I'm not certain whether your shared logic will correctly pull out the details it needs from nested node pools.  My read is that it will, but I'm glad you're writing a test so we can be sure.

darrenhaken commented on google/regional_utils.go:
```diff
@@ -0,0 +1,37 @@
+//These functions are used by both the `resource_container_node_pool` and `resource_container_cluster` for handling regional clusters
+
+func isZone(location string) bool {
+	return len(strings.Split(location, "-")) == 3
```
I didn't think some of these functions were particularly robust, if I get time I can look to use Regex instead. If I do that I'll add unit tests for the function too.

darrenhaken commented:
@ndmckinley after seeing the prefix logic I thought you might say that the shared functions may not work. 

If I extend them to support a prefix will they then work?

ndmckinley commented:
I can't say for certain right now, but I think so.  Can you point me at which shared functions you are looking at in particular?

darrenhaken commented:
I meant the regional_utils.go functions.

ndmckinley commented:
I think that stuff should all work fine with nested node pools.  We'll find out for sure when we've got a test case.

ndmckinley commented:
Once you add in another test that does the update I was talking about in that PR comment above, I'll be ready to merge it.  I'm running the tests now - as long as they all pass, plus this new one, we can get this in today.  I recall your timezone and mine diverge an awful lot - I'll be available and watching the thread til around 6:30 PST.

darrenhaken commented:
@ndmckinley there's a failing test right now which I'm trying to work out. Any help you can offer would be appreciated.

The test is `TestAccContainerNodePool_autoscaling` in the node pool tests.

Output I get is:

```
--- FAIL: TestAccContainerNodePool_autoscaling (782.23s)
	testing.go:513: Step 2 error: Error applying: 1 error(s) occurred:

		* google_container_node_pool.np: 1 error(s) occurred:

		* google_container_node_pool.np: googleapi: Error 400: Invalid JSON payload received. Unknown name "update": Cannot find field., badRequest
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	782.263s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-google/scripts	0.026s [no tests to run]
make: *** [testacc] Error 1
```

I did some debugging and the error is raised on https://github.com/darrenhaken/terraform-provider-google/blob/master/google/resource_container_node_pool.go#L540

ndmckinley commented on google/resource_container_node_pool.go:
```diff
@@ -155,11 +151,57 @@ var schemaNodePool = map[string]*schema.Schema{
+	return &NodePoolInformation{
+		project:  project,
+		location: location,
+		cluster:  d.Get("cluster").(string),
```
This is giving panics:

Edited for space:
```
panic: interface conversion: interface {} is nil, not string

github.com/terraform-providers/terraform-provider-google/google.extractNodePoolInformation(0xc42030ae00, 0xc420658000, 0x0, 0xc42010d160, 0x0)
resource_container_node_pool.go:198 +0x262
github.com/terraform-providers/terraform-provider-google/google.nodePoolUpdate(0xc42030ae00, 0x18126a0, 0xc420658000, 0xc4202ee600, 0x17, 0xc420530970, 0xc, 0xa, 0xc42041b680, 0x1a)
resource_container_node_pool.go:507 +0x68
github.com/terraform-providers/terraform-provider-google/google.resourceContainerClusterUpdate(0xc42030ae00, 0x18126a0, 0xc420658000, 0x0, 0xc4207dce08)
resource_container_cluster.go:1136 +0x74c
```

darrenhaken replied:
Is that from running the test suite or related to the autoscaling test I mentioned?

ndmckinley replied:
From running the test suite.  Same panic in 

`TestAccContainerCluster_withAdditionalZones`
`TestAccContainerCluster_withAddons`
`TestAccContainerCluster_withLegacyAbac`
`TestAccContainerCluster_withMasterAuthorizedNetworksConfig`

Possibly others - tests aren't done yet.

I'm looking at that test you mentioned as well.

darrenhaken replied:
OK, I can take a look at that. I wonder if it's because the `cluster` key requires a prefix to resolve in the map?

ndmckinley replied:
I'm betting that's it.

darrenhaken replied:
I've passed along the prefix to it now and I'm running `TestAccContainerCluster_withAddons` so I'll see what the output is.

If you can provide any insight into the autoscaling test above that will be awesome 👍 

I'll try my best to do some work this evening for the good of regionals!

darrenhaken replied:
It didn't work adding the prefix, attaching the debugger to see if I can see other issues.

darrenhaken replied:
So it is the cluster key. That key does not exist from the perspective from the perspective of a cluster, I'm looking at at fix now

darrenhaken replied:
@ndmckinley I have pushed a fix tested it with `TestAccContainerCluster_withAddons` and it's passing.

See commit for info.

darrenhaken commented:
@ndmckinley to answer your point about needing a test that also completes an update. It looks like `TestAccContainerNodePool_autoscaling` already does that. Will that satisfy the update requirement?

ndmckinley commented:
Re: update tests.  I might be overcautious, but container clusters are super complicated, and I don't know much about the differences between how a regional cluster handles updates and how a regular one does, so I was hoping to see a test almost exactly like `_autoscaling` but with a regional cluster.

As for that `autoscaling` test, I'm trying to run it locally to capture the incorrect request and figure it out.

darrenhaken commented:
@ndmckinley OK I look to add that other test in...

I think it's fair to say that clusters and node pools are very complicated. Long term I reckon it will be handy to try and extract some of the shared logic and put some unit tests around it. A job for another day :)

darrenhaken commented:
@ndmckinley fixed the issues you raised (hopefully) with `TestAccContainerCluster_withAddons`. Still not too sure about the Update issue I reported and I might not get much more time to work on it tonight.

ndmckinley commented:
I'll let you know if those tests pass.

ndmckinley commented:
Aha, figured out what went wrong with your update.  You sent the request for a [Cluster update](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/update) to a [NodePool](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters.nodePools/update).

```
PUT /v1beta1/projects/nmckinley-graphite/locations/us-central1-a/clusters/tf-nodepool-test-8mbc73xx6o/nodePools/tf-nodepool-test-sma28uf3f3?alt=json HTTP/1.1 
...
{"update":{"desiredNodePoolAutoscaling":{"enabled":true,"maxNodeCount":5,"minNodeCount":0},"desiredNodePoolId":"tf-nodepool-test-sma28uf3f3"}}
```

See how it's got the `nodePools` bit in the URL there?

ndmckinley commented:
Running it now, sure thing.

darrenhaken commented:
Thanks. Keep me posted and I'll try my best to respond and fix things.

Would love to have this resolved tonight so I can have some closure and a
regional cluster :-)

On Thu, 19 Apr 2018, 19:05 Nathan McKinley, <notifications@github.com>
wrote:

> Running it now, sure thing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-382828786>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me32WZ6U5fIxa33dJVFjzYPwJgRcLks5tqNHlgaJpZM4TQb8T>
> .
>

ndmckinley commented:
Three failures:

```
------- Stdout: -------
=== RUN   TestAccContainerCluster_withIPAllocationPolicy
--- FAIL: TestAccContainerCluster_withIPAllocationPolicy (877.90s)
    testing.go:506: Step 3, expected error:
        
        Error applying: 1 error(s) occurred:
        
        * google_container_cluster.with_ip_allocation_policy: 1 error(s) occurred:
        
        * google_container_cluster.with_ip_allocation_policy: Error waiting for creating GKE cluster: secondary range "services" does not exist in network "container-net-4kx2raa6ko", subnetwork "container-net-4kx2raa6ko".
        
        To match:
        
        services secondary range "services" not found in subnet
```

```------- Stdout: -------
=== RUN   TestAccContainerCluster_withNetworkPolicyEnabled
--- FAIL: TestAccContainerCluster_withNetworkPolicyEnabled (613.87s)
    testing.go:513: Step 4 error: Check failed: Check 1/1 error: google_container_cluster.with_network_policy_enabled: Attribute 'network_policy.0.enabled' not found
FAIL
```

And a handful of quota issues I'll try to get resolved.

darrenhaken commented:
Does `TestAccContainerNodePool_withGPU` pass for you @ndmckinley? That looked like it definitely shouldn't do.

ndmckinley commented:
I didn't run that test this time around - just the ContainerCluster tests.  But no, I don't think it passes.

darrenhaken commented:
`with_ip_allocation_policy` failing seems odd because its nothing to do with node pools.

I'm sifting through the codebase, it's tricky without any line of code to narrow the search.
I appreciate any help.

darrenhaken commented:
@ndmckinley I am clueless to why this is still not working. Is there any chance you could help?

It looks like the only function used from the node pool resource by the cluster resource is `nodePoolUpdate`. I attached a debugger to the function and ran `TestAccContainerCluster_withNetworkPolicyEnabled`. It doesn't hit the breakpoint.

ndmckinley commented:
Let me see what I can find out, but I can't promise to commit too much more time to this this week.  Early next week I'll be a bit more free.

darrenhaken commented:
Ok thanks I appreciate it. I'll try and see what I can do to help.

On Fri, 20 Apr 2018, 19:04 Nathan McKinley, <notifications@github.com>
wrote:

> Let me see what I can find out, but I can't promise to commit too much
> more time to this this week. Early next week I'll be a bit more free.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-383176904>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6MexJGQ13WfPvohz7GzZxB_39yb1Ivks5tqiNHgaJpZM4TQb8T>
> .
>

ndmckinley commented:
Hm, I've come to believe that `withNetworkPolicyEnabled` is not your fault at all.  I see it failing at head now and I might have a fix for it.  Let me run the rest of the tests on your PR and I'll stop harassing you about that one (and `withIPAllocationPolicy`, which is also failing at head).

ndmckinley commented:
The only tests I'm seeing broken and new are things that I think have been fixed in master since you started this PR - can you merge master into the PR so I can be sure?

darrenhaken commented:
Yes I'll do it now

On Fri, 20 Apr 2018, 19:37 Nathan McKinley, <notifications@github.com>
wrote:

> The only tests I'm seeing broken and new are things that I think have been
> fixed in master since you started this PR - can you merge master into the
> PR so I can be sure?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-383185518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me60J89eMH-sDTL9TF0_olZ6pc6rjks5tqirmgaJpZM4TQb8T>
> .
>

darrenhaken commented:
Right I've merged it, is it up to date?

ndmckinley commented:
Think so!  The tests are pretty slow, but expect to hear back from me in a little while.

darrenhaken commented:
OK, please keep me posted.

ndmckinley commented:
Unfortunately there are still two tests which pass at HEAD but fail on your PR.

```
------- Stdout: -------
=== RUN   TestAccContainerNodePool_version
--- FAIL: TestAccContainerNodePool_version (529.54s)
    testing.go:513: Step 2 error: Error applying: 1 error(s) occurred:
        
        * google_container_node_pool.np: 1 error(s) occurred:
        
        * google_container_node_pool.np: googleapi: Error 400: Node pool id must be specified if cluster has more than one node pool.., badRequest
FAIL
```

And

```
------- Stdout: -------
=== RUN   TestAccContainerNodePool_withGPU
--- FAIL: TestAccContainerNodePool_withGPU (518.11s)
    testing.go:513: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        DESTROY/CREATE: google_container_node_pool.np_with_gpu
          cluster:                                 "tf-cluster-nodepool-test-izwhv0l1gn" => "tf-cluster-nodepool-test-izwhv0l1gn"
 [ ... things that are the same removed ... ]
          node_config.0.preemptible:               "true" => "true"
          node_config.0.service_account:           "default" => "default"
          node_config.0.taint.#:                   "1" => "0" (forces new resource)
          node_config.0.taint.0.effect:            "NO_SCHEDULE" => "" (forces new resource)
          node_config.0.taint.0.key:               "nvidia.com/gpu" => "" (forces new resource)
          node_config.0.taint.0.value:             "present" => "" (forces new resource)
          node_count:                              "1" => "<computed>"
          project:                                 "hc-terraform-testing" => "<computed>"
```

darrenhaken commented:
When I run the GPU test on my master it fails, I can't see how it could pass when the version of K8 is not available anymore. It needs upgrading on the test to 1.9.6. Are you sure it's passing?

I can look into the other test

darrenhaken commented:
@ndmckinley I have fixed the version test but I am still having problems with the GPU test. In fact, I have ran it twice and it failed for two different reasons (different to yours).

Failure 1:
This was a huge output but the error seemed to be
```
 Dependencies:
		    google_container_cluster.cluster
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error applying: 1 error(s) occurred:

		* google_container_cluster.cluster (destroy): 1 error(s) occurred:

		* google_container_cluster.cluster: Error waiting for deleting GKE cluster: Get https://container.googleapis.com/v1/projects/at-delivery-platform-testing/zones/us-central1-c/operations/operation-1524293483699-0eacafbd?alt=json: read tcp 192.168.0.13:55611->216.58.204.74:443: read: connection reset by peer
```

Failure 2:
```
--- FAIL: TestAccContainerNodePool_withGPU (166.39s)
	testing.go:513: Step 0 error: Error applying: 1 error(s) occurred:

		* google_container_node_pool.np_with_gpu: 1 error(s) occurred:

		* google_container_node_pool.np_with_gpu: Error waiting for creating GKE NodePool: Get https://container.googleapis.com/v1beta1/projects/at-delivery-platform-testing/locations/us-central1-c/operations/operation-1524294184725-e2fbc5ef?alt=json: read tcp 192.168.0.13:55727->216.58.204.74:443: read: connection reset by peer
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error refreshing: 1 error(s) occurred:

		* google_container_cluster.cluster: google_container_cluster.cluster: Error reading instance group manager returned as an instance group URL: "Get https://www.googleapis.com/compute/beta/projects/at-delivery-platform-testing/zones/us-central1-c/instanceGroupManagers/gke-tf-cluster-nodepool--default-pool-d3886709-grp?alt=json: read tcp 192.168.0.13:55740->216.58.204.74:443: read: connection reset by peer"

		State: <nil>
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	166.417s
```

darrenhaken commented:
@ndmckinley quick update, I have pushed another change which has improved the reliability of the GPU test and I am now able to reproduce the error you reported each time consistently.

However, I cannot work out what the issue is and I think I need some assistance.

darrenhaken commented:
@ndmckinley do you think you'll get some time today to help look?

darrenhaken commented:
@kri5 if you fancy taking a look at the failing `TestAccContainerNodePool_withGPU` I'd be more than happy to take input ;-)

ndmckinley commented:
Hi, @darrenhaken, sorry, haven't had time to take a look yet, but still might get to it today.  If not today, I'll do my best to make sure it's tomorrow.

darrenhaken commented:
Hey, @ndmckinley is today looking like a better day?

ndmckinley commented:
Yep, I'm on it.  I'll see what I can find out.

darrenhaken commented:
Thanks I appreciate it. I took another look today but I'm still none the
wiser.

On Tue, 24 Apr 2018, 18:55 Nathan McKinley, <notifications@github.com>
wrote:

> Yep, I'm on it. I'll see what I can find out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-384023564>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6MezyymSyXIe7qjAwVxuq841TNuJA2ks5tr2cXgaJpZM4TQb8T>
> .
>

Stono commented:
By me.
Also, I feel the need to let you know it's my birthday. And all I want for
my birthday is a working regional cluster with node pools provisioned by
terraform.

On Tue, 24 Apr 2018, 7:23 pm Darren Haken, <notifications@github.com> wrote:

> Just got schooled on taint, ignore that :-)
>
> On Tue, 24 Apr 2018, 19:21 Darren Haken, <darrenhaken@gmail.com> wrote:
>
> > What's a taint?
> >
> > Ok great if you need anything let me know, we are honestly really keen to
> > get this so we can create our cluster on GCP.
> >
> > We will appreciate it :-)
> >
> > On Tue, 24 Apr 2018, 19:19 Nathan McKinley, <notifications@github.com>
> > wrote:
> >
> >> Okay, I see what's happening - the taint that gets added to make the
> >> cluster get created with GPUs is getting removed, somehow. Let me find
> out
> >> why and come back to you with better details.
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly, view it on GitHub
> >> <
> https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-384031147
> >,
> >> or mute the thread
> >> <
> https://github.com/notifications/unsubscribe-auth/AA6Me4gVguffcjBfngrpMc9N9eMRxdFpks5tr2y8gaJpZM4TQb8T
> >
> >> .
> >>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-384032246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaviYsgJy5agSfRtRW7Xtz0miF-GPx4ks5tr22dgaJpZM4TQb8T>
> .
>

ndmckinley commented:
Happy birthday.  I'll see what we can do.  :)

ndmckinley commented:
You'll want to look at the `DiffSuppressFunc` here: https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go and you'll want to apply that to the taint field in `node_config.go`, I _think_. I'm only pretty sure that's a good solution... Mind giving it a try?

darrenhaken commented:
Trying a run through of the test with `DiffSuppressFunc`

darrenhaken commented:
@ndmckinley The above function didn't work with the `DiffSuppressFunc`.
I might struggle to do much more tonight (getting late here), would you be able to carry on? 
I'll make you a contributor on the repo just in case you want to commit anything.

One other idea I had is adding back in the old v1 API calls for zonal clusters, would that fix the issue?
This would mean introducing the switch based on region.

ndmckinley commented:
Sure, I'll give it a shot.

darrenhaken commented:
Awesome, I only ask because it's a birthday...

On Tue, 24 Apr 2018, 21:46 Nathan McKinley, <notifications@github.com>
wrote:

> Sure, I'll give it a shot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-384074489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Me1yg53xVnl0JeBTQZWCIwFsWSF_kks5tr489gaJpZM4TQb8T>
> .
>

ndmckinley commented:
:)  I've proven to myself that this is definitely 100% the issue, and I have a fix that _works_, but is too overbroad.  I'll get it down to the minimum possible interference and send it to your branch there.

Assuming it's all right with you if I merge the PR once that's done?

darrenhaken commented:
Absolutely and awesome on proving it! If you fix and merge I'll be
incredibly happy :-)

Will a release be quick to follow on?

On Tue, 24 Apr 2018, 22:42 Nathan McKinley, <notifications@github.com>
wrote:

> :) I've proven to myself that this is definitely 100% the issue, and I
> have a fix that *works*, but is too overbroad. I'll get it down to the
> minimum possible interference and send it to your branch there.
>
> Assuming it's all right with you if I merge the PR once that's done?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-google/pull/1320#issuecomment-384089454>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6MezpxjgoVyG-qXfzpo-8ndMSxXdu7ks5tr5w_gaJpZM4TQb8T>
> .
>

ndmckinley commented:
Tests running now.  GPU test passes locally ... fingers crossed!

ndmckinley commented:
All right, I've got one thing which I think is a spurious test failure ... I'm re-running it.

ndmckinley commented:
Good news - the failed test passes on a retry, as do all the others.  It's going in!  Happy birthday, @Stono.  😆

morgante commented:
Yay! So happy to see this merged—it's not my birthday, but it might as well be. 🎆

darrenhaken commented:
@ndmckinley awesome work! Thanks so much for helping me get over the final line with it. When can we expect it to be released?

Stono commented:
IM SO HAPPY I COULD CRY.
@ndmckinley we had brownies to celebrate.  I ate yours for you

herrewig commented:
Thank you!!!!!! :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2: :hearts: :star2:

ndmckinley commented:
@darrenhaken I see the questions around release!  Our releases are usually on a biweekly schedule, and last release was last Friday.

AndrewFarley commented:
Heck yes!  Just built the provider myself to test, I really needed this, and it works great.  Thanks!!!  +1 +1 +1
